config/config.go

```diff
@@ -0,0 +1,209 @@
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/encoding"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+var (
+	// DefaultZoneConfig is the default zone configuration
+	// used when no custom config has been specified.
+	DefaultZoneConfig = &ZoneConfig{
+		ReplicaAttrs: []proto.Attributes{
+			{},
+			{},
+			{},
+		},
+		RangeMinBytes: 1048576,
+		RangeMaxBytes: 67108864,
+		GC: &GCPolicy{
+			TTLSeconds: 24 * 60 * 60, // 1 day
+		},
+	}
+
+	// ZoneConfigHook is a function used to lookup a zone config given a table
+	// or database ID.
+	// This is also used by testing to simplify fake configs.
+	ZoneConfigHook func(*SystemConfig, uint32) (*ZoneConfig, error)
+
+	// TestingLargestIDHook is a function used to bypass GetLargestObjectID
+	// in tests.
+	TestingLargestIDHook func() uint32
+
+	// TestingDisableTableSplits is a testing-only variable that disables
+	// splits of tables into separate ranges.
+	TestingDisableTableSplits bool
+)
+
+// ObjectIDForKey returns the object ID (table or database) for 'key',
+// or (_, false) if not within the structured key space.
+func ObjectIDForKey(key proto.Key) (uint32, bool) {
+	if key.Equal(proto.KeyMax) {
+		return 0, false
+	}
+	if key.Equal(keys.TableDataPrefix) {
+		// TODO(marc): this should eventually return SystemDatabaseID.
+		return 0, false
+	}
+	remaining := bytes.TrimPrefix(key, keys.TableDataPrefix)
+	if len(remaining) == len(key) {
+		// TrimPrefix returns the input untouched if the prefix doesn't match.
+		return 0, false
+	}
+	// Consume first encoded int.
+	// TODO(marc): we really need to be able to handle non-Uvarint keys.
+	_, id64 := encoding.DecodeUvarint(remaining)
+	return uint32(id64), true
+}
+
+// Get searches the kv list for 'key' and returns its
+// raw byte value if found. ok is true only if the key is found.
+func (s *SystemConfig) Get(key proto.Key) ([]byte, bool) {
+	l := len(s.Values)
+	index := sort.Search(l, func(i int) bool {
+		return bytes.Compare(s.Values[i].Key, key) >= 0
+	})
+	if index == l {
+		return nil, false
+	}
+	// TODO(marc): I'm pretty sure a Value returned by MVCCScan can
+	// never be nil. Should check.
+	return s.Values[index].Value.Bytes, true
+}
+
+// GetLargestObjectID returns the largest object ID found in the config.
+// This could be either a table or a database.
+func (s *SystemConfig) GetLargestObjectID() (uint32, error) {
+	if TestingLargestIDHook != nil {
+		return TestingLargestIDHook(), nil
+	}
+
+	if len(s.Values) == 0 {
+		return 0, fmt.Errorf("empty system values in config")
+	}
+
+	// descriptorTableID = 3
+	// Search for the first key after the descriptor table.
+	// We can't use Get as we don't mind if there is nothing after
+	// the descriptor table.
```

Comment:
This comment is confusing. I think what you're doing is retrieving the last key in the descriptor table and you're doing that because that descriptor may have no keys associated with it.